### PR TITLE
Added boundary flux and boundary source

### PR DESCRIPTION
**Pull request by mattlie82**
Added `bflux` and `bsource` to describe surface fluxes and (fixed) sources. 

The example in `Example311_HeatEquation_BoundaryDiffusion.jl` still shows problems (artefacts near boundaries).
Moreover, it is currently not possible to describe anything but homogeneous Neumann boundary conditions for the surface fluxes. This needs additionally functionality in `ExtendableGrids.jl`.

### CODE REVIEW

**Comment by mattlie82** [comment]
Tests fail due to allocations during assembly in boundary face loop.

File: examples/Example311_HeatEquation_BoundaryDiffusion.jl [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        bflux       = function(f, u0, edge)
+                        u        = unknowns(edge,u0)
+                        f[2]     = eps_surf*(u[2,1]-u[2,2])
+                      end,
+        breaction   = function(f,u,node)
+                        if node.region == breg
+                          f[1] = k*(u[1]-u[2])
+                          f[2] = k*(u[2]-u[1])
+                        else
+                          f[1] = 0
```
Comment (j-fu) [inline examples/Example311_HeatEquation_BoundaryDiffusion.jl:62]: Use f[1]=0.0  for type stability here. May well be the reason of the allocations.

**Review by j-fu — CHANGES_REQUESTED** [review]
I tend to agree with some whitespace chanegs, but these are hard to review together with functionality changes. Let us separate them in the future. One big formatting update would be better, for this JuliaFormatter.jl might be the way to go.

As for the allocations: I'd like to see all tests running. You could try to switch checking off when testing locally:  https://github.com/j-fu/VoronoiFVM.jl/blob/master/test/runtests.jl#L5

File: examples/Example311_HeatEquation_BoundaryDiffusion.jl (hunk 1 above)
Comment (mattlie82) [reply examples/Example311_HeatEquation_BoundaryDiffusion.jl:62]: IT appears that the allocations happen in `eval_and_assemble` in the loop over the boundary edges when filling the edge element:
```
  edge.index   = bfaceedges[ibedge, ibface]
  edge.node[1] = bedgenodes[1, edge.index]
  edge.node[2] = bedgenodes[2, edge.index]
```

File: src/vfvm_solver.jl [hunk 2]
```diff
@@ -266,97 +266,105 @@ function eval_and_assemble(system::AbstractSystem{Tv, Ti},
...
-        edgenodes=grid[EdgeNodes]
-        has_celledges=true
+        cellx         = grid[CellEdges]
+        edgenodes     = grid[EdgeNodes]
+        has_celledges = true
     end
 
-    nn::Int=num_nodes(geom)
-    ne::Int=num_edges(geom)
+    bedgenodes = grid[BEdgeNodes]
```
Comment (j-fu) [inline src/vfvm_solver.jl:353]: We need a type annotation here. In fact I need to check with ExtendableSparse why this is necessary at all. May be @chmerdon has a similar problem.

```
bedgenodes::Array{Ti,2}=...
```

Comment (chmerdon) [reply src/vfvm_solver.jl:353]: Yes, I once encountered similar problems when accessing grid component, type annotations usually help in these situations.

Comment (mattlie82) [reply src/vfvm_solver.jl:353]: Same here, now it runs.

**Comment by j-fu** [comment]
Ok, kann es sein dass Du ein modifiziertes ExtendableGrids  nutzt ? Dann bräuchte erst einmal den PR dort.

**Comment by mattlie82** [comment]
> Ok, kann es sein dass Du ein modifiziertes ExtendableGrids nutzt ? Dann bräuchte erst einmal den PR dort.
I did, see https://github.com/j-fu/ExtendableGrids.jl/pull/8
